### ISSUE TITLE
Why emplace enterpoint_copy while epDeleted in addPoint?

### ISSUE THREAD

**Issue by jinwenabc** [nmslib/hnswlib#308]
in the function of addPoint, after searching the top candidates of the layers below curlevel, it then checks if enter_point deleted. If deleted, emplace the enterpoint_copy into top candidates? I'm not clear about the purpose of emplacing the enterpoint_copy, could some one give some hint? thanks

**Comment by yurymalkov** [comment]
Hm. I am not sure. It probably allows to avoid degradation of performance due to not updating connections of the enterpoint, but looking back it seems like the right thing to do is to completely ignore the remove flag during addition, as we expect the deleted points can be brought back to life.
So, yeah, this needs to be changed.

**Comment by jinwenabc** [comment]
Yes, ignoring the flag may be better. btw, if a point is marked as deleted, the connections of that point should be kept?
And if a deleted node is searched during a search operation, it can be filtered. If so, are there any options to periodically clear some deleted points? or clear the deleted points once arrive at a certain amount/ratio?

**Comment by yurymalkov** [comment]
There is an option to rebuilt the index from scratch (if the number of deleted elements is large), or update the deleted elements to new elements, so that they the graph nodes can be reused (though this is a costly operation). 
The problem with deleting the nodes is that is hard to do locally, since the connections that lead to a deleted node are not specifically indexed (thus one would need to brute force all index or miss some connections to the deleted element, leading to checks which decrease the performance). Updating/reusing the nodes allows to avoid this problem.
